File: app/src/main/java/com/narcyber/mvpbasics/presenter/RegisterActivityPresenter.java
package com.narcyber.mvpbasics.presenter;

import android.content.Context;

import androidx.annotation.NonNull;

import com.narcyber.mvpbasics.helper.DataSaveHelper;
import com.narcyber.mvpbasics.model.User;

public class RegisterActivityPresenter extends ParentPresenter {

    private final RegisterView view;
    private final DataSaveHelper<User> dataSaveHelper;

    public RegisterActivityPresenter(RegisterView view, Context context) {
        super();
        this.view = view;
        dataSaveHelper = new DataSaveHelper<User>(context);
    }

    public boolean isUsernameTaken(@NonNull String username) {

        getUserRepository().requestUserByUsername(username);

        return true;
    }

    @Override
    public void respondUserByUserName(User user) {
        if (user != null) {
            view.isUsernameUsed(true);
            return;
        }
        view.isUsernameUsed(false);
    }


    public void removeUser(String key) {
        dataSaveHelper.removeObject(key);
    }

    public boolean isEmailTaken(@NonNull String email) {
        getUserRepository().requestUserByEmail(email);
        return true;
    }

    @Override
    public void respondUserByEmail(User user) {
        if (user != null) {
            view.isEmailUsed(true);
            return;
        }
        view.isEmailUsed(false);
    }

    public void pushUserIntoDb(final String userEmail, final String userFullName,
                               final String userPassword, final String userName) {
        final User user = new User(userEmail, userFullName, userName, userPassword);
        getUserRepository().requestUserPush(user);
    }

    @Override
    public void respondSuccessPushing() {
        view.notifyUserSuccessRegistered();
    }

    @Override
    public void respondPushingFailed(String message) {
        view.notifyUserRegFailed();
    }

    public interface RegisterView {

        boolean isEmailUsed(boolean isUsed);

        boolean isUsernameUsed(boolean isUsed);

        boolean notifyUserSuccessRegistered();

        boolean notifyUserRegFailed();
    }

}

File: app/src/main/java/com/narcyber/mvpbasics/presenter/RegisterActivityPresenter.kt
package com.narcyber.mvpbasics.presenter

import com.narcyber.mvpbasics.model.User

class RegisterActivityPresenter(val view: RegisterView) :
    ParentPresenter() {
    fun isUserNameTaken(userName: String = ""): Boolean {
        userRepository.requestUserByUsername(userName)
        return true
    }

    fun isEmailTaken(email: String = ""): Boolean {
        userRepository.requestUserByEmail(email)
        return true
    }

    fun pushUserIntoDB(
        email: String,
        userName: String,
        name: String,
        password: String,
    ) {
        val user = User(email = email, fullName = name, userName = userName, password = password)
        userRepository.requestUserPush(user)
    }

    override fun respondSuccessPushing() {
        view.notifyUserSuccessRegistered()
    }

    override fun respondPushingFailed(message: String?) {
        view.notifyUserRegFailed()
    }

    override fun respondUserByEmail(user: User?) {
        if (user == null) {
            view.isEmailUsed(false)
            return
        }
        view.isEmailUsed(true)
    }

    override fun respondUserByUserName(user: User?) {
        if (user == null) {
            view.isUsernameUsed(false)
            return
        }
        view.isUsernameUsed(true)
    }

    interface RegisterView {
        fun isEmailUsed(isUsed: Boolean): Boolean
        fun isUsernameUsed(isUsed: Boolean): Boolean
        fun notifyUserSuccessRegistered(): Boolean
        fun notifyUserRegFailed(): Boolean
    }

}